File: Automation/src/test/java/homework4/TestLocatorsUsage.java
package homework4;

import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;
import org.testng.Assert;
import org.testng.annotations.Test;

public class TestLocatorsUsage extends LocatorUsageBaseClass{
    @Test
    void testLocators() {
        driver.get(BaseUrl+"upload-download");
        WebElement searchPart = driver.findElement(By.cssSelector(".col-12.mt-4.col-md-6"));
        WebElement toCompare = driver.findElement(By.xpath("//*[@id=\"downloadButton\"]"));
        WebElement currentElement;

        currentElement = searchPart.findElement(By.id("downloadButton"));
        Assert.assertEquals(toCompare, currentElement);
        currentElement = searchPart.findElement(By.className("btn-primary"));
        Assert.assertEquals(toCompare, currentElement);
        currentElement = searchPart.findElement(By.tagName("a"));
        Assert.assertEquals(toCompare, currentElement);
        currentElement = searchPart.findElement(By.linkText("Download"));
        Assert.assertEquals(toCompare, currentElement);
        currentElement = searchPart.findElement(By.partialLinkText("Downl"));
        Assert.assertEquals(toCompare, currentElement);
        currentElement = driver.findElement(By.cssSelector("div.col-12.mt-4.col-md-6 a[download='sampleFile.jpeg']"));
        Assert.assertEquals(toCompare, currentElement);
    }
}

File: homework4/TestLocatorsUsage.py
import pytest
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.remote.webelement import WebElement


@pytest.mark.usefixtures("chrome_driver")
class TestLocatorsUsage:
    driver: webdriver.Chrome
    BASEURL = "https://demoqa.com/"

    def test_locators_usage(self):
        self.driver.get(self.BASEURL+"upload-download")

        search_part: WebElement = self.driver.find_element(By.CSS_SELECTOR(".col-12.mt-4.col-md-6"))
        to_compare: WebElement = self.driver.find_element(By.XPATH("//*[@id=\"downloadButton\"]"))
        current_element: WebElement

        current_element = search_part.find_element(By.ID("downloadButton"))
        assert to_compare == current_element

        current_element = search_part.find_element(By.CLASS_NAME("btn-primary"))
        assert to_compare == current_element

        current_element = search_part.find_element(By.TAG_NAME("a"))
        assert to_compare == current_element

        current_element = search_part.find_element(By.LINK_TEXT("Download"))
        assert to_compare == current_element

        current_element = search_part.find_element(By.PARTIAL_LINK_TEXT("Downl"))
        assert to_compare == current_element

        current_element = self.driver.find_element(By.CSS_SELECTOR("div.col-12.mt-4.col-md-6 a[download='sampleFile.jpeg']"))
        assert to_compare == current_element

